plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -212,17 +213,123 @@ processTerm = handleTerm <=< traverseOf termSubtypes applyTypeSubstitution where
-            -- inlining. Because `inlineSaturatedApp` traverses *all* application nodes for each
-            -- subterm, the runtime is quadratic for terms with a long chain of applications.
-            -- If we use the context-based approach like in GHC, this won't be a problem, so we may
-            -- consider that in the future.
-            inlineSaturatedApp =<< forMOf termSubterms t processTerm
+        t -> do
+            -- See note [Processing order of call site inlining]
+            let (hd, args) = splitApplication t
+                -- process the term args, ignore the type arguments
+                processArgs :: AppContext tyname name uni fun ann
+                    -> InlineM tyname name uni fun ann (AppContext tyname name uni fun ann)
+                processArgs (TermAppContext arg ann ctx) = do
+                    processedArg <- processTerm arg
+                    processedArgs <- processArgs ctx
+                    pure $ TermAppContext processedArg ann processedArgs
+                processArgs (TypeAppContext ty ann ctx) = do
+                    processedArgs <- processArgs ctx
+                    pure $ TypeAppContext ty ann processedArgs
+                processArgs AppContextEnd = pure AppContextEnd
+            case (hd, args) of
+                (Var{}, _) -> do
+                    processedHd <- processTerm hd -- the variable may be unconditionally inlined.
+                    case processedHd of
+                        -- if it didn't get unconditionally inlined/got inlined to another variable
+                        (Var _ name) -> do
+                                gets (lookupVarInfo name) >>= \case
+                                    Just varInfo -> do
+                                        -- process the args
+                                        processedArgs <- processArgs args
+                                        maybeInlined <- callSiteInline t varInfo processedArgs
+                                        case maybeInlined of
+                                            Just inlined -> pure inlined
+                                            -- we didn't inline at the call site, just process the
+                                            -- subterms
+                                            Nothing      -> forMOf termSubterms t processTerm
+                                    -- The variable maybe a *recursive* let binding, in which case
+                                    -- it won't be in the map, and we don't process it.
+                                    -- ATM recursive bindings aren't inlined.
+                                    Nothing -> forMOf termSubterms t processTerm
+                        -- if the processed head is not a variable (because it's inlined), just
+                        -- process the subterms
+                        _ ->  forMOf termSubterms t processTerm
```

Comment:
Here the head of `t` is already processed (i.e., `processedHd`), and `forMof termSubterms ...` will process it again. May not be a problem but I'm curious what @michaelpj thinks.

Reply:
I'm indeed a bit confused. This looks rather different to what Marty and I came up with when we paired on it and I'm not sure why? I was expecting something like:
```
case args of
  -- not really an application, so hd is the term itself. Processing it will loop.
  -- I feel like we came up with a nicer way of handling this but I've forgotten
  AppContextEnd -> forMOf termSubterms t processTerm
  _ -> do 
    hd' <- processTerm head
    args' <- processArgs args
    let reconstructed = fillAppCtx hd' args'
    case hd' of
      (Var _name) -> do
           ...
           case maybeInlined oof
              Just inlined -> pure inlined
              Nothing -> pure reconstructed
      _ -> pure reconstructed
```
Here the logic is very clear: if we split it into an application, then we process the parts of the application and put it together again. We don't do the subterm traversal an additional time.

Reply:
After I paired with you, Ziyang wanted me to only split the context when it's a variable. And then there are still the bugs. This result is after I modified that to get rid of the errors.

Reply:
> only split the context when it's a variable

Doing so minimizes the changes to the `processTerm` logic, but I think Michael's way is better - it's shorter and I can't think of any issue it may cause. Can you try it? I'd be surprised if it causes any errors or regression.

Reply:
yes, Michael paired with me and we tried that, and got the errors. Yes I know we got surprised by the inliner over and over again.

Reply:
> yes, Michael paired with me and we tried that, and got the errors.

This is surprising since it seems correct to me. Can you push a commit that does what Michael suggested? I just want to see what happens.

Reply:
https://github.com/input-output-hk/plutus/pull/5426/commits/caf0d825d2f959c9351f55df7acee1171338a948

Reply:
I have since fixed a few things since that commit though. But yeah you can see the idea.

Reply:
> I have since fixed a few things

That might have fixed the problem. Just put caf0d82 on top of your branch pls - I just want to see how it does.

Reply:
There will be merge conflict, I can't "just" put that on top. In fact, I tried that and it's just too messy. I put the fixes I did on top of [this branch](https://github.com/input-output-hk/plutus/tree/thealmarty/michaels-approach-pir-inliner). When I ran it locally I got 
```
GHC Core to PLC plugin: Error: Error from the PIR compiler:
                               Error from the PLC compiler:
                               Free type variable at  Ann {annInline = MayInline, annSrcSpans = { no-src-span }} :  a_10327
Context: Compiling expr at "plutus-benchmark-0.1.0.0-inplace-script-contexts-internal:PlutusBenchmark.ScriptContexts:(96,7)-(96,53)"
Error: cabal: Failed to build lib:script-contexts-internal from
plutus-benchmark-0.1.0.0 (which is required by
test:plutus-benchmark-script-contexts-tests from plutus-benchmark-0.1.0.0).
```
Let me know if you get the same result.

(I don't want to spend a lot of time tidying this branch up so I only put the semantic relevant fixes in)